doc/developer/design/20230531_compute_metrics.md

```diff
@@ -0,0 +1,414 @@
+  * [ ] `mz_compute_response_message_size_bytes`
+    * **Type**: histogram
+    * **Labels**: `instance_id`, `replica_id`
+    * **Description**: A histogram of sizes of compute response messages, by replica.
+    * **Notes**: This metric exists already as `mz_compute_messages_received_bytes`.
+                 Proposing to rename it because "messages received" doesn't imply who the receiver is.
+* Controller state
+  * [ ] `mz_compute_controller_replica_count`
+    * **Type**: gauge
+    * **Labels**: `instance_id`
+    * **Description**: The number of replicas.
+  * [ ] `mz_compute_controller_collection_count`
+    * **Type**: gauge
+    * **Labels**: `instance_id`
+    * **Description**: The number of installed compute collections.
+  * [ ] `mz_compute_controller_peek_count`
+    * **Type**: gauge
+    * **Labels**: `instance_id`
+    * **Description**: The number of pending peeks.
+  * [ ] `mz_compute_controller_subscribe_count`
+    * **Type**: gauge
+    * **Labels**: `instance_id`
+    * **Description**: The number of active subscribes.
+  * [ ] `mz_compute_controller_command_queue_size`
+    * **Type**: gauge
+    * **Labels**: `instance_id`, `replica_id`
+    * **Description**: The size of the compute command queue, by replica.
+  * [ ] `mz_compute_controller_response_queue_size`
+    * **Type**: gauge
+    * **Labels**: `instance_id`, `replica_id`
+    * **Description**: The size of the compute response queue, by replica.
+* Command history
+  * [ ] `mz_compute_controller_history_command_count`
+    * **Type**: gauge
+    * **Labels**: `instance_id`, `type`
+    * **Description**: The number of commands in the command history, by command type.
+  * [ ] `mz_compute_controller_history_dataflow_count`
+    * **Type**: gauge
+    * **Labels**: `instance_id`
+    * **Description**: The number of dataflows in the command history.
+* Replica connection state
+  * [ ] `mz_compute_controller_connected_replica_count`
+    * **Type**: gauge
+    * **Labels**: `instance_id`
+    * **Description**: The number of successfully connected replicas.
+  * [ ] `mz_compute_controller_replica_connects_total`
+    * **Type**: counter
+    * **Labels**: `instance_id`, `replica_id`
+    * **Description**: The total number of replica (re-)connections, by replica.
+  * [ ] `mz_compute_controller_replica_connect_wait_time_seconds_total`
+    * **Type**: counter
+    * **Labels**: `instance_id`, `replica_id`
+    * **Description**: The total time spent waiting for replica (re-)connection, by replica.
+* Peeks
+  * [ ] `mz_peeks_total`
+    * **Type**: counter
+    * **Labels**: `instance_id`, `result`
+    * **Description**: The total number of peeks served, by result type (success, error, canceled).
+  * [ ] `mz_peek_duration_seconds`
+    * **Type**: histogram
+    * **Labels**: `instance_id`, `result`
+    * **Description**: A histogram of peek durations since restart, by result type (success, error, canceled).
```

Comment:
For purposes of standardization and discoverability, would it be possible to prefix all of these metrics with `mz_compute_`, or would this be misleading in some cases?

Reply:
For `mz_peeks_*` I don't have strong feelings. Peeks are a compute concept, so there shouldn't be any ambiguity without "compute", but discoverability is probably a good reason to still have a prefix.

The other metrics that lack a "compute" prefix are the `mz_dataflow` and `mz_arrangement` metrics. For future-proofing reasons I think we shouldn't mention compute here. Once we have cluster unification we will hopefully be able to include storage dataflow in these metrics.